File: go/quantum_safe_signatures.go
// Quantum Safe - Digital Signatures
package pqc

import (
	"github.com/cloudflare/circl/sign/dilithium/mode2"
	"github.com/cloudflare/circl/sign/dilithium/mode3"
	"github.com/cloudflare/circl/sign/ed448"
	"github.com/open-quantum-safe/liboqs-go/oqs"
)

func MLDsa44Signature() {
	pk, sk, _ := mode2.GenerateKey(nil)
	message := []byte("quantum safe message")
	signature := mode2.Sign(sk, message)
	valid := mode2.Verify(pk, message, signature)
	_ = valid
}

func MLDsa65Example() {
	publicKey, secretKey, _ := mode3.GenerateKey(nil)
	sig := mode3.Sign(secretKey, []byte("test data"))
	_ = publicKey
	_ = sig
}

func Falcon512Signature() {
	signer := oqs.Signature{}
	signer.Init("Falcon-512", nil)
	pk, _ := signer.GenerateKeyPair()
	msg := []byte("falcon signed message")
	sig, _ := signer.Sign(msg)
	valid, _ := signer.Verify(msg, sig, pk)
	_ = valid
}

func SLHDsaSphincsPlus() {
	signer := oqs.Signature{}
	signer.Init("SPHINCS+-SHA256-128f-robust", nil)
	pk, _ := signer.GenerateKeyPair()
	signature, _ := signer.Sign([]byte("data"))
	_ = pk
	_ = signature
}

func DilithiumSignature() {
	scheme := mode3.Scheme()
	pk, sk, _ := scheme.GenerateKey()
	sig := scheme.Sign(sk, []byte("dilithium message"), nil)
	_ = pk
	_ = sig
}

func XMSSSignatureExample() {
	signer := oqs.Signature{}
	signer.Init("Dilithium2", nil)
	pk, _ := signer.GenerateKeyPair()
	signature, _ := signer.Sign([]byte("xmss message"))
	_ = pk
	_ = signature
}

func LMSHashBasedSignature() {
	// LMS - Leighton-Micali Signature
	signer := oqs.Signature{}
	signer.Init("Dilithium3", nil)
	_, _ = signer.GenerateKeyPair()
}

File: rust/quantum_safe_signatures.rs
// Quantum Safe - Digital Signatures
use pqcrypto_dilithium::dilithium2;
use pqcrypto_dilithium::dilithium3;
use pqcrypto_falcon::falcon512;
use pqcrypto_sphincsplus::sphincsshake256128frobust;
use oqs::sig::{Sig, Algorithm as SigAlgorithm};

fn ml_dsa_44_signature() {
    let (pk, sk) = dilithium2::keypair();
    let message = b"quantum safe message";
    let signature = dilithium2::sign(message, &sk);
    let verified = dilithium2::verify(message, &signature, &pk);
}

fn ml_dsa_65_example() {
    let (public_key, secret_key) = dilithium3::keypair();
    let sig = dilithium3::sign(b"test data", &secret_key);
}

fn falcon_512_signature() {
    let (pk, sk) = falcon512::keypair();
    let msg = b"falcon signed message";
    let sig = falcon512::sign(msg, &sk);
    falcon512::verify(msg, &sig, &pk).unwrap();
}

fn slh_dsa_sphincs_plus() {
    let (pk, sk) = sphincsshake256128frobust::keypair();
    let signature = sphincsshake256128frobust::sign(b"data", &sk);
}

fn xmss_signature_example() {
    let sig = Sig::new(SigAlgorithm::Dilithium2).unwrap();
    let (pk, sk) = sig.keypair().unwrap();
    let signature = sig.sign(b"xmss message", &sk).unwrap();
}

fn lms_hash_based_signature() {
    // LMS - Leighton-Micali Signature
    let lms_params = LmsParams::new(LmsAlgorithm::LMS_SHA256_M32_H10);
    let keypair = lms_keygen(&lms_params);
}
